components/service/glean/src/main/java/mozilla/components/service/glean/config/Configuration.kt

```diff
@@ -14,45 +14,47 @@ import mozilla.components.service.glean.net.PingUploader
  *
  * @property serverEndpoint the server pings are sent to. Please note that this is
  *           is only meant to be changed for tests.
+ * @property channel the release channel the application is on, if known. This will be
+ *           sent along with all the pings, in the `client_info` section.
  * @property userAgent the user agent used when sending pings, only to be used internally.
  * @property maxEvents the number of events to store before the events ping is sent
  * @property logPings whether to log ping contents to the console. This is only meant to be used
  *           internally by the `GleanDebugActivity`.
  * @property httpClient The HTTP client implementation to use for uploading pings.
  * @property pingTag String tag to be applied to headers when uploading pings for debug view.
  *           This is only meant to be used internally by the `GleanDebugActivity`.
- * @property channel the release channel the application is on, if known. This will be
- *           sent along with all the pings, in the `client_info` section. */
+ */
 data class Configuration internal constructor(
     val serverEndpoint: String,
+    val channel: String? = null,
     val userAgent: String = DEFAULT_USER_AGENT,
     val maxEvents: Int = DEFAULT_MAX_EVENTS,
     val logPings: Boolean = DEFAULT_LOG_PINGS,
     // NOTE: since only simple object or strings can be made `const val`s, if the
     // default values for the lines below are ever changed, they are required
     // to change in the public constructor below.
     val httpClient: PingUploader = ConceptFetchHttpUploader(lazy { HttpURLConnectionClient() }),
-    val pingTag: String? = null,
-    val channel: String? = null
+    val pingTag: String? = null
 ) {
     // This is the only public constructor this class should have. It should only
     // expose things we want to allow external applications to change. Every test
     // only or internal configuration option should be added to the above primary internal
     // constructor and only initialized with a proper default when calling the primary
     // constructor from the secondary, public one, below.
+    @JvmOverloads
     constructor(
         serverEndpoint: String = DEFAULT_TELEMETRY_ENDPOINT,
+        channel: String? = null,
         maxEvents: Int = DEFAULT_MAX_EVENTS,
-        httpClient: PingUploader = ConceptFetchHttpUploader(lazy { HttpURLConnectionClient() }),
-        channel: String? = null
+        httpClient: PingUploader = ConceptFetchHttpUploader(lazy { HttpURLConnectionClient() })
```

Comment:
what's the reason for the change of order here?

Reply:
The `JvmOverloads` is basically generating as many function as the number of defaults. So we'll have:

```java
Configuration(server)
Configuration(server, maxEvents)
Configuration(server, maxEvents, httpClient)
Configuration(server, maxEvents, httpClient, channel)
```

So, with the previous ordering, in order to set the channel, we'd need to provide *all* the arguments, including the `httpClient`. This means we'd need to Javify `ConceptFetchHttpUploader` as well. If we did that, to set the channel, we'd need to:

```java
Configuration("someserver", 200, new ConceptFetchHttpUpdate(new Kotlin.lazy(.....)), "release")
```

vs

```kotlin
Configuration().copy(channel="release)
```

in kotlin. We could probably consider making the `channel` the first parameter, since it's the one that would most likely change.

Reply:
Just to confirm: hopefully that doesn't break the API for existing (Kotlin) usages?

Reply:
Not that I found, no. We use named arguments in [Fenix](https://github.com/mozilla-mobile/fenix/blob/8a9340fd9d3ab4e5d1f849cd1de8bd3a2bde4a78/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L402)/reference-browser. So we should be fine.